specification/dartLangSpec.tex

```diff
@@ -2530,91 +2530,169 @@ \section{Classes}
-a concrete accessible member with the same name as $m$,
-with a method signature $m'$ which is not a correct override of $m$
+
+\LMHash{}%
+Each concrete member must have a suitable signature:
+Assume that $C$ has a concrete member with
+the same name as $m$ and accessible to $L$,
+and let \DefineSymbol{m''} be its member signature.
+\commentary{%
+The concrete member may be declared in $C$ or inherited from a superclass.%
+}
+Let \DefineSymbol{m'} be the member signature which is obtained from $m''$
+by adding, if not present already, the modifier \COVARIANT{}
+(\ref{covariantParameters})
+to each parameter $p$ in $m''$ where
+the corresponding parameter in $m$ has the modifier \COVARIANT.
+It is a compile-time error if $m'$ is not a correct override of $m$
 (\ref{correctMemberOverrides}),
 unless that concrete member is a \code{noSuchMethod} forwarder
 (\ref{theMethodNoSuchMethod}).
 
 \commentary{%
-In particular, it is an error for a class to be concrete even if it inherits
-a member implementation for every member signature in its interface,
-unless each of them has parameters and types such that they satisfy
-the corresponding member signature.
-But when there is a non-trivial \code{noSuchMethod} it is allowed
-to leave some members unimplemented,
-and it is allowed to have a \code{noSuchMethod} forwarder which does not
-satisfy the class interface
-(in which case it will be overridden by another \code{noSuchMethod} forwarder).%
+Consider a concrete class \code{C},
+and assume that \code{C} declares or inherits
+a member implementation with the right name
+for every member signature in its interface.
+It is still an error if one or more of them has parameters or types
+such that they do not satisfy
+the corresponding member signature in the interface.
+For this check, any missing \COVARIANT{} modifiers are implicitly added
+to the signature of an inherited member (this is how we get $m'$ from $m''$).
+When the modifier \COVARIANT{} is added to one or more parameters
+(which will only happen when the concrete member is inherited),
+an implementation may choose to implicitly induce a forwarding method
+with the same signature as $m'$,
+in order to perform the required dynamic type check,
+and then invoke the inherited method.%
 }
 
 \LMHash{}%
-Assume that $C$ has a concrete accessible member with the same name as $m$,
-with a method signature $m'$ which is a correct override of $m$.
-For each parameter $p$ of $m'$ where \COVARIANT{} is present,
-it is a compile-time error if there exists
-a direct or indirect superinterface of $C$ which has
-an accessible method signature $m''$ with the same name as $m$,
-such that $m''$ has a parameter $p''$ that corresponds to $p$
-(\ref{covariantParameters}),
-unless the type of $p$ is a subtype or a supertype of the type of $p''$.
+The use of an implicitly induced forwarding method
+when the modifier \COVARIANT{} is added to
+one or more parameters in $m'$
+is allowed,
```

Comment:
"is allowed" is intended to mean that a conforming implementation can do it, but it could also use some other implementation technique. I've changed it to use "implementation specific", that's better.

> it can then get run-time failures on seemingly
> type-valid invocations

We must have the dynamic checks in any case, because those upcasts on the receiver can make a seemingly type-valid invocation incur a type error at run time.

> the forwarder is then one implementation approach
> to desugaring that behavior, but it's visible at runtime
> whether you have a forwarder or not on the runtime type
> of the tearoff

Exactly.
When we say that it is allowed for an implementation to use the forwarding method, we avoid requiring that the inherited method (somehow) performs the dynamic type checks on its actual arguments. This could yield better performance, because we might otherwise perform those dynamic type checks in a lot of method implementations